learningequality/studio
# Channel history tracking and prober endpoint for publishing status alerts

## Summary
### Background
- In the last few months we've seen several forum posts regarding channels that became "stuck" (no forward progress) during publishing
- As release manager, I would like to avoid interrupting channel publication as much as possible. It's possible that initiating a release deploy can interrupt the publish process which requires manual correction. Checking the status of channel publishing or to manually resolve a channel publishing issue requires production shell access, something that I would like to minimize.
- As release manager, I would like to be alerted when a channel's publication becomes stalled for any reason or takes too long so that I can take immediate intervention. This includes channel publication interrupted by a release deploy, an infrastructural issue slowing down task processing, restart/redeploy of server infrastructure, or other extant bugs.
- As a contributor interested in security, I would like to ensure that we have auditing in place to give us insight into potential security problems should they arise. This auditing can help fill gaps that we've uncovered in auditing channel permissions, such as the initial creator of a channel.

### Description of the change(s) you made

- Adds new model `ChannelHistory` to associate history of creation, publication, deletion, and recovery of channels with user who initiated it and when
- Updates channel procedures to create relevant `ChannelHistory`
- Adds prober endpoint for returning publishing status, utilizing `ChannelHistory`

### Manual verification steps performed
1. Manually update `export_channel_task` to `time.sleep(10000)`
2. Publish a channel
3. As admin, load `/api/probers/publishing_status`
4. Ensure that the channel is listed in the endpoint response
5. Run the prober from the command line: `PROBER_PUBLISHING_ALERT_THRESHOLD=1 python deploy/probers/publishing_status_probe.py`

### Screenshots (if applicable)


### Does this introduce any tech-debt items?

- When to prune history
___
## Reviewer guidance
### How can a reviewer test these changes?



### Are there any risky areas that deserve extra testing?



## References


## Comments


----

### Contributor's Checklist


PR process:


Studio-specifc:



Testing:

___

### Reviewer's Checklist
#### This section is for reviewers to fill out.
